src/main/java/net/sf/jsqlparser/statement/select/CNFConverter.java

```diff
@@ -0,0 +1,614 @@
+ *                     OR(M)
+ *           /     /         \              \
+ *          F     H          NOT            AND(M)
+ *                            |      /       |       \
+ *                            G    NOT      OR(M)    OR(M)
+ *                                  |      /    \   /    \
+ *                                  H     J     L  K      L 
+ * 
+ * Now let us push the and up and we will get the result like this:
+ *                             AND(M)
+ *             /                |                 \
+ *            OR(M)             OR(M)             OR(M)
+ *     /    /   \   \    /   /  |   \   \    /  /  |   \  \
+ *    F    H    NOT NOT  F  H  NOT  J   L   F  H  NOT  K   L
+ *               |   |          |                  |
+ *               G   H          G                  G
+ *               
+ * 5. The last step, convert the Multiple Expression back to the binary
+ * form. Note the final tree shall be left-inclined. 
+ * 
+ * The final expression tree shall be like this:
+ *                                            AND
+ *                                       /                \
+ *                                   AND                ( )
+ *                             /            \            |
+ *                           ( )            ( )         part1
+ *                            |              |
+ *                            OR          part2
+ *                        /        \
+ *                       OR        NOT
+ *                  /         \     |
+ *                OR          NOT   H
+ *            /        \       |
+ *           F          H      G
+ *     
+ * part1:                                         OR
+ *                                          /            \
+ *                                         OR            L
+ *                                    /          \
+ *                                  OR            K
+ *                              /        \
+ *                            OR          NOT    
+ *                         /      \        |
+ *                        F        H       G
+ * 
+ * part2:                                        OR
+ *                                         /        \
+ *                                        OR          L
+ *                                   /          \
+ *                                 OR            J
+ *                            /          \
+ *                          OR           NOT
+ *                      /        \        |                            
+ *                     F          H       G                                    
+ * 
+ * @author messfish
+ *
+ */
+public class CNFConverter {
+
+    private Expression root; 
+    // the variable that stores the newly generated root.
+    private Expression dummy;
+    // this variable mainly serves as the dummy root of the true root.
+    // generally it will be a multi and operator with root as the child.
+    private Expression temp1;
```

Comment:
Due to this variables your class will not be threadsafe. You should prevent double usage, shoudnt you?